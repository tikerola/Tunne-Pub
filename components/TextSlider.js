import React, { useState } from 'react';
import {
  FlatList,
  SafeAreaView,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
  Dimensions
} from 'react-native';

import { useDispatch, useSelector } from 'react-redux'


export default function TextSlider({ textData, saveToRedux, question, columns, wordType }) {

  const words = useSelector(state => state.cards.summaryOfChosenItems[state.cards.activeSummaryItemId][wordType])
  const dispatch = useDispatch()
  

  const toggleSelectedWords = word => {
    const wordExists = words.includes(word)
    
    if (wordExists) {
      dispatch(saveToRedux(wordType, word, false, question))
    }
    else {
      dispatch(saveToRedux(wordType, word, true, question))
    }
  }

  return (
    <SafeAreaView style={styles.container}>
      <Text style={styles.info}>*{wordType === 'sentences' ? 'Lauselista' : 'Sanalista'} jatkuu vierittämällä näyttöä</Text>
      <FlatList
        testID="flatlist"
        style={styles.flatlist}
        data={textData}
        numColumns={columns}
        extraData={words}
        keyExtractor={item => item.id}
        initialNumToRender={20}
        renderItem={({ item }) => <TouchableOpacity
          onPress={() => {
            toggleSelectedWords(item.word)
          }}
          style={[styles.wordContainer, { transform: [{ scale: words.includes(item.word) ? 1.1 : 1 }], width: columns === 2 ? '50%' : '100%' }]}
        >
          <View>
            <Text style={[styles.word, words.includes(item.word) && { color: 'red', transform: [{ translateX: 3 }] }]}>{item.word}</Text>
          </View>
        </TouchableOpacity>}
      />
    </SafeAreaView>
  )
}


const styles = StyleSheet.create({
  container: {
    width: '100%',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
  },

  info: {
    width: '100%',
    marginTop: 10,
    marginLeft: '15%',
    marginBottom: 10,
    fontSize: 16,
    fontFamily: 'Roboto-Light',
    color: '#999'
  },

  flatlist: {
    marginTop: 10,
    width: '95%',
    height: Dimensions.get('screen').height * 0.55,
    marginBottom: 15
  },

  wordContainer: {
    padding: 5,
    marginTop: 5,
    marginBottom: 5,
    marginLeft: 0,
    marginRight: 0,


    alignItems: 'center'
  },

  word: {
    fontSize: 18,
    fontFamily: 'Segoe-Print',
    color: '#bbb'
  }
});

