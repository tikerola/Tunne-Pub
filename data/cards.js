import Card from "../models/card";
import * as images from "./images";

const CARD_DATA = [
  new Card("1", "happy", images.image1),
  new Card("2", "sad", images.image2),
  new Card("3", "worried", images.image3),
  new Card("4", "scared", images.image4),
  new Card("5", "happy", images.image5),
  new Card("6", "sad", images.image6),
  new Card("7", "worried", images.image7),
  new Card("8", "scared", images.image8),

  new Card("9", "happy", images.image9),
  new Card("10", "sad", images.image10),
  new Card("11", "worried", images.image11),
  new Card("12", "scared", images.image12),
  new Card("13", "happy", images.image13),
  new Card("14", "sad", images.image14),
  new Card("15", "worried", images.image15),
  new Card("16", "scared", images.image16),

  new Card("17", "happy", images.image17),
  new Card("18", "sad", images.image18),
  new Card("19", "worried", images.image19),
  new Card("20", "scared", images.image20),
  new Card("21", "happy", images.image21),
  new Card("22", "sad", images.image22),
  new Card("23", "worried", images.image23),
  new Card("24", "scared", images.image24),

  new Card("25", "happy", images.image25),
  new Card("26", "sad", images.image26),
  new Card("27", "worried", images.image27),
  new Card("28", "scared", images.image28),
  new Card("29", "happy", images.image29),
  new Card("30", "sad", images.image30),
  new Card("31", "worried", images.image31),
  new Card("32", "scared", images.image32),

  new Card("33", "happy", images.image33),
  new Card("34", "sad", images.image34),
  new Card("35", "worried", images.image35),
  new Card("36", "scared", images.image36),
  new Card("37", "happy", images.image37),
  new Card("38", "sad", images.image38),
  new Card("39", "worried", images.image39),
  new Card("40", "scared", images.image40),

  new Card("41", "happy", images.image41),
  new Card("42", "sad", images.image42),
  new Card("43", "worried", images.image43),
  new Card("44", "scared", images.image44),
  new Card("45", "happy", images.image45),
  new Card("46", "sad", images.image46),
  new Card("47", "worried", images.image47),
  new Card("48", "scared", images.image48),
  new Card("49", "scared", images.image49),
];

export default CARD_DATA;
