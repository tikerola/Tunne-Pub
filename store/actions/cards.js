export const TOGGLE_CHOSEN_CARD = "TOGGLE CHOSEN CARD";
export const ADD_SUMMARY_ITEM = "ADD SUMMARY ITEM";
export const DELETE_SUMMARY_ITEM = "DELETE SUMMARY ITEM";
export const ADD_ANSWER_TO_A_QUESTION = "ADD ANSWER TO A QUESTION";
export const LOAD_SUMMARY_ITEMS = "LOAD SUMMARY ITEMS";

export const SET_ACTIVE_SUMMARY_ITEM_ID = "SET ACTIVE SUMMARY ITEM ID";
export const SET_SCREEN_NUMBER = "SET SCREEN NUMBER";
export const ADD_SUMMARY_ITEM_NAME = "ADD SUMMARY ITEM NAME";
export const ITEM_SAVED = "ITEM SAVED";
export const SET_NUM_OF_ITEMS = "SET NUM OF ITEMS";
export const TOGGLE_WORD = "TOGGLE WORD";

export const toggleWord = (typeOfWord, word, isAdd, question) => ({
  type: TOGGLE_WORD,
  typeOfWord,
  payload: word,
  isAdd,
  question,
});

export const setNumOfItems = (num) => {
  return {
    type: SET_NUM_OF_ITEMS,
    num,
  };
};

export const itemSaved = (id) => {
  return {
    type: ITEM_SAVED,
    id,
  };
};

export const setScreenNumber = (screenNumber) => {
  return {
    type: SET_SCREEN_NUMBER,
    screenNumber,
  };
};

export const setActiveSummaryItemId = (id) => ({
  type: SET_ACTIVE_SUMMARY_ITEM_ID,
  id,
});

export const toggleChosenCard = (card, isAdd) => {
  return {
    type: TOGGLE_CHOSEN_CARD,
    payload: card,
    isAdd,
  };
};

export const addSummaryItem = (id) => {
  return {
    type: ADD_SUMMARY_ITEM,
    id,
  };
};

export const addSummaryItemName = (id, name) => {
  return {
    type: ADD_SUMMARY_ITEM_NAME,
    id,
    name,
  };
};

export const deleteSummaryItem = (id) => {
  return {
    type: DELETE_SUMMARY_ITEM,
    id,
  };
};

export const addAnswerToAQuestion = (
  summaryItemId,
  question,
  answer,
  listMode,
  wordType
) => {
  return {
    type: ADD_ANSWER_TO_A_QUESTION,
    id: summaryItemId,
    question,
    answer,
    listMode,
    wordType,
  };
};

export const loadSummaryItems = (items) => {
  return {
    type: LOAD_SUMMARY_ITEMS,
    items,
  };
};
