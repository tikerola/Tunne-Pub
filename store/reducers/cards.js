import CART_DATA from "../../data/cards";
import WORD_DATA from "../../data/words";
import NEEDS_DATA from "../../data/needs";
import { QUESTIONS } from "../../data/questions";

import {
  TOGGLE_CHOSEN_CARD,
  ADD_SUMMARY_ITEM,
  DELETE_SUMMARY_ITEM,
  ADD_ANSWER_TO_A_QUESTION,
  LOAD_SUMMARY_ITEMS,
  SET_ACTIVE_SUMMARY_ITEM_ID,
  SET_SCREEN_NUMBER,
  ADD_SUMMARY_ITEM_NAME,
  ITEM_SAVED,
  SET_NUM_OF_ITEMS,
  TOGGLE_WORD,
} from "../actions/cards";
import { SENTENCES_DATA } from "../../data/sentences";

const initialState = {
  cards: CART_DATA,
  words: WORD_DATA,
  needs: NEEDS_DATA,
  sentences: SENTENCES_DATA,
  activeSummaryItemId: "",
  summaryOfChosenItems: {},
  screenNumber: 1,
  numOfItems: 0,
  initialLoadFinished: false,
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_NUM_OF_ITEMS:
      return {
        ...state,
        numOfItems: action.num,
      };

    case ITEM_SAVED: {
      const summaryItems = { ...state.summaryOfChosenItems };
      summaryItems[action.id].itemSaved = true;

      return {
        ...state,
        summaryOfChosenItems: summaryItems,
      };
    }

    case ADD_SUMMARY_ITEM_NAME: {
      const summaryItems = { ...state.summaryOfChosenItems };
      summaryItems[action.id].name = action.name;

      return {
        ...state,
        summaryOfChosenItems: summaryItems,
      };
    }

    case SET_SCREEN_NUMBER:
      return {
        ...state,
        screenNumber: action.screenNumber,
      };

    case SET_ACTIVE_SUMMARY_ITEM_ID:
      return {
        ...state,
        activeSummaryItemId: action.id,
      };

    case TOGGLE_CHOSEN_CARD: {
      const itemToEdit = {
        ...state.summaryOfChosenItems[state.activeSummaryItemId],
      };

      if (action.isAdd)
        itemToEdit.cards = [...itemToEdit.cards, action.payload];
      else
        itemToEdit.cards = itemToEdit.cards.filter(
          (card) => card.id !== action.payload.id
        );

      return {
        ...state,
        summaryOfChosenItems: {
          ...state.summaryOfChosenItems,
          [state.activeSummaryItemId]: itemToEdit,
        },
      };
    }

    case TOGGLE_WORD: {
      const itemToEdit = {
        ...state.summaryOfChosenItems[state.activeSummaryItemId],
      };

      if (action.isAdd) {
        itemToEdit[action.typeOfWord] = [
          ...itemToEdit[action.typeOfWord],
          action.payload,
        ];
      } else
        itemToEdit[action.typeOfWord] = itemToEdit[action.typeOfWord].filter(
          (word) => word !== action.payload
        );

      if (action.typeOfWord === "sentences")
        itemToEdit.questions[QUESTIONS[action.question]] = itemToEdit[
          action.typeOfWord
        ].join(" ");
      else
        itemToEdit.questions[QUESTIONS[action.question]] = itemToEdit[
          action.typeOfWord
        ].join(", ");

      return {
        ...state,
        summaryOfChosenItems: {
          ...state.summaryOfChosenItems,
          [state.activeSummaryItemId]: itemToEdit,
        },
      };
    }

    case ADD_SUMMARY_ITEM:
      return {
        ...state,
        activeSummaryItemId: action.id,
        summaryOfChosenItems: {
          ...state.summaryOfChosenItems,
          [action.id]: {
            id: action.id,
            name: "",
            cards: [],
            feelingsWords: [],
            feelingsYouMissWords: [],
            needs: [],
            needsBehindAFeeling: [],
            sentences: [],
            questions: {},
            itemSaved: false,
          },
        },
      };

    case DELETE_SUMMARY_ITEM:
      const newSummaryItems = { ...state.summaryOfChosenItems };
      delete newSummaryItems[action.id];
      return {
        ...state,
        activeSummaryItemId:
          state.activeSummaryItemId === action.id
            ? ""
            : state.activeSummaryItemId,
        summaryOfChosenItems: newSummaryItems,
      };

    case ADD_ANSWER_TO_A_QUESTION:
      const newSummaryItem = { ...state.summaryOfChosenItems[action.id] };
      newSummaryItem.questions[action.question] = action.answer;

      if (!action.listMode && action.wordType) {
        newSummaryItem[action.wordType] = [];
      }

      return {
        ...state,
        summaryOfChosenItems: {
          ...state.summaryOfChosenItems,
          [action.id]: newSummaryItem,
        },
      };

    case LOAD_SUMMARY_ITEMS:
      return {
        ...state,
        initialLoadFinished: true,
        summaryOfChosenItems: action.items,
        numOfItems: Object.keys(action.items).length,
      };

    default:
      return state;
  }
};

export default reducer;
