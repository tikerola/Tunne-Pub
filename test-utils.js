import React from "react";
import { NavigationContext } from "@react-navigation/native";
import { render } from "@testing-library/react-native";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import CARD_DATA from "./data/cards";
import NEEDS_DATA from "./data/needs";
import { SENTENCES_DATA } from "./data/sentences";
import WORD_DATA from "./data/words";

const INITIAL_STATE = {
  cards: {
    // cards: CARD_DATA,
    // words: WORD_DATA,
    // needs: NEEDS_DATA,
    // sentences: SENTENCES_DATA,
    activeSummaryItemId: "123",
    summaryOfChosenItems: {
      123: {
        id: "123",
        name: "",
        cards: [],
        feelingsWords: [],
        feelingsYouMissWords: [],
        needs: ["hei"],
        needsBehindAFeeling: [],
        sentences: [],
        questions: {},
        itemSaved: true,
      },
    },
    screenNumber: 1,
    numOfItems: 0,
    initialLoadFinished: false,
  },
};

const mockStore = configureStore([]);
const store = mockStore(INITIAL_STATE);

const navContext = {
  isFocused: () => true,
  // addListener returns an unscubscribe function.
  addListener: jest.fn(() => jest.fn()),
};

const reduxRender = (ui, options, storeData = store) =>
  render(ui, {
    wrapper: ({ children }) => (
      <Provider store={storeData}>
        <NavigationContext.Provider value={navContext}>
          {children}
        </NavigationContext.Provider>
      </Provider>
    ),
    ...options,
  });

// re-export everything for convenience
export * from "@testing-library/react-native";

// override render method
export { reduxRender as render, store };
