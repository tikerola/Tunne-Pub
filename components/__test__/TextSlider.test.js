
import React from 'react'
import { TouchableOpacity } from 'react-native-gesture-handler'
import { render, fireEvent, store, waitFor, screen } from '../../test-utils'
import TextSlider from '../TextSlider'
import { toggleWord } from '../../store/actions/cards'

afterAll(() => {
    store.clearActions()
})


describe('<TextSlider />', () => {

    const textData = [
        {
            id: '1',
            word: 'moi'
        },
        {
            id: '2',
            word: 'hei'
        },
        {
            id: '3',
            word: 'jippii'
        }
    ]

    

    it('should match snapshot', () => {
        const result = render(<TextSlider />).toJSON()
        expect(result).toMatchSnapshot()
    })
    
    it('should trigger an action with right parameters depending on whats stored in redux', async () => {
        const rendered = render(<TextSlider textData={textData} saveToRedux={toggleWord} wordType="needs" />)
        
        let items = rendered.UNSAFE_getAllByType('Text')

        fireEvent.press(items[1])
        let actions = await waitFor(() => store.getActions())

        expect(actions[0].type).toEqual('TOGGLE WORD')
        expect(actions[0].isAdd).toEqual(true)
        expect(actions[0].typeOfWord).toEqual('needs')

        fireEvent.press(items[2])
        actions = await waitFor(() => store.getActions())
        expect(actions[1].isAdd).toEqual(false)
        
    })

})