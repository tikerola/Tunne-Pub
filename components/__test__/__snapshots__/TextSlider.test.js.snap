// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextSlider /> should match snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#999",
        "fontFamily": "Roboto-Light",
        "fontSize": 16,
        "marginBottom": 10,
        "marginLeft": "15%",
        "marginTop": 10,
        "width": "100%",
      }
    }
  >
    *
    Sanalista
     jatkuu vierittämällä näyttöä
  </Text>
  <RCTScrollView
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={20}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "height": 733.7,
        "marginBottom": 15,
        "marginTop": 10,
        "width": "95%",
      }
    }
    testID="flatlist"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</RCTSafeAreaView>
`;
